### PR TITLE
1 4 task.valery.aleshka

### PR DESCRIPTION
**Pull request by Valeryaleshka**

### CODE REVIEW

**Comment by Valeryaleshka** [comment]
https://travis-ci.org/github/Valeryaleshka/js-assignments/builds/715660855

File: task/01-strings-tasks.js [hunk 1]
```diff
@@ -22,7 +22,10 @@
  *   '',  'bb'  => 'bb'
  */
 function concatenateStrings(value1, value2) {
-    throw new Error('Not implemented');
+
+    return value1+value2;
```
Comment (KostyaKov) [inline task/01-strings-tasks.js]: It is necessary to respect the indentation between operators
https://javascript.info/coding-style

File: task/01-strings-tasks.js [hunk 2]
```diff
@@ -22,7 +22,10 @@
...
+
+    //throw new Error('Not implemented');
```
Comment (KostyaKov) [inline task/01-strings-tasks.js]: Comments of unnecessary code are redundant

File: task/01-strings-tasks.js [hunk 3]
```diff
@@ -38,7 +41,10 @@ function concatenateStrings(value1, value2) {
  *   ''      => 0
  */
 function getStringLength(value) {
-    throw new Error('Not implemented');
+
```
Comment (KostyaKov) [inline task/01-strings-tasks.js:42]: Blank lines in this case are redundant

File: task/01-strings-tasks.js [hunk 4]
```diff
@@ -69,7 +78,11 @@ function getStringFromTemplate(firstName, lastName) {
  *   'Hello, Chuck Norris!' => 'Chuck Norris'
  */
 function extractNameFromTemplate(value) {
-    throw new Error('Not implemented');
+
+    let re = /Hello, (\w+)\s(\w+)!/;
+    return value.replace(re, "$1 $2");
```
Comment (KostyaKov) [inline task/01-strings-tasks.js]: A simpler solution would be to use empty string replacement with empty string

File: task/01-strings-tasks.js [hunk 5]
```diff
@@ -114,7 +133,14 @@ function removeLeadingAndTrailingWhitespaces(value) {
  *   'cat', 3 => 'catcatcat'
  */
 function repeatString(value, count) {
-    throw new Error('Not implemented');
+
+    let string ="";
+    for(let i = 0; i<count; i++){
```
Comment (KostyaKov) [inline task/01-strings-tasks.js]: For this task you can use "repeat" function

File: task/01-strings-tasks.js [hunk 6]
```diff
@@ -130,7 +156,14 @@ function repeatString(value, count) {
  *   'ABABAB','BA' => 'ABAB'
  */
 function removeFirstOccurrences(str, value) {
-    throw new Error('Not implemented');
+
+    let index = str.indexOf(value);
```
Comment (KostyaKov) [inline task/01-strings-tasks.js]: For this task you can use just "replace" function

File: task/01-strings-tasks.js [hunk 7]
```diff
@@ -221,7 +307,10 @@ function getRectangleString(width, height) {
  *
  */
 function encodeToRot13(str) {
-    throw new Error('Not implemented');
+
+    return str.replace(/[a-zA-Z]/g,function(c){return String.fromCharCode((c<="Z"?90:122)>=(c=c.charCodeAt(0)+13)?c:c-26);});
```
Comment (KostyaKov) [inline task/01-strings-tasks.js]: It is necessary to respect the indentation between operators

File: task/01-strings-tasks.js [hunk 8]
```diff
@@ -238,7 +327,14 @@ function encodeToRot13(str) {
  *   isString(new String('test')) => true
  */
 function isString(value) {
-    throw new Error('Not implemented');
+
+    if (typeof value === "string"|| value instanceof String){
```
Comment (KostyaKov) [inline task/01-strings-tasks.js]: Here you can return result of the expression without "if"

File: task/01-strings-tasks.js [hunk 9]
```diff
@@ -267,7 +363,15 @@ function isString(value) {
  *   'K♠' => 51
  */
 function getCardId(value) {
-    throw new Error('Not implemented');
+
+    let cards = ['A♣','2♣','3♣','4♣','5♣','6♣','7♣','8♣','9♣','10♣','J♣','Q♣','K♣',
```
Comment (KostyaKov) [inline task/01-strings-tasks.js]: Must use "const" for this case

File: task/02-numbers-tasks.js [hunk 10]
```diff
@@ -89,7 +101,14 @@ function getDistanceBetweenPoints(x1, y1, x2, y2) {
  *   5*x = 0         => 0
  */
 function getLinearEquationRoot(a, b) {
-    throw new Error('Not implemented');
+
+    let result = -b/a;
+    if(result===0){
```
Comment (KostyaKov) [inline task/02-numbers-tasks.js]: This task can be performed in one "return" expression without unnecessary checking

File: task/02-numbers-tasks.js [hunk 11]
```diff
@@ -202,7 +246,23 @@ function roundToPowerOfTen(num, pow) {
  *   17 => true
  */
 function isPrime(n) {
-    throw new Error('Not implemented');
+
+    if (n === 1) {
+        return false;
+    }
+    else if (n === 2) {
```
Comment (KostyaKov) [inline task/02-numbers-tasks.js]: "else" after "return" is redundant

File: task/04-arrays-tasks.js [hunk 12]
```diff
@@ -304,7 +366,10 @@ function get3TopItems(arr) {
  *   [ 1, '2' ] => 1
  */
 function getPositivesCount(arr) {
-   throw new Error('Not implemented');
+
+   return arr.filter(v => Number.isInteger(v)).filter(v => v>0).length;
```
Comment (KostyaKov) [inline task/04-arrays-tasks.js]: Here you can do with one use of "filter" with two conditions inside

File: task/04-arrays-tasks.js [hunk 13]
```diff
@@ -353,7 +427,13 @@ function getItemsSum(arr) {
  *  [ null, undefined, NaN, false, 0, '' ]  => 6
  */
 function getFalsyValuesCount(arr) {
-   throw new Error('Not implemented');
+
+   let arr2 = arr.filter(function (n) {
+      return (n === false || n === 0 || n === "" || n === '' || n === null || n === undefined || n !== n)
```
Comment (KostyaKov) [inline task/04-arrays-tasks.js]: Here you can use simply "(item => !item)"

File: task/04-arrays-tasks.js [hunk 14]
```diff
@@ -536,7 +670,11 @@ function selectMany(arr, childrenSelector) {
  *   [[[ 1, 2, 3]]], [ 0, 0, 1 ]      => 2        (arr[0][0][1])
  */
 function getElementByIndexes(arr, indexes) {
-    throw new Error('Not implemented');
+   
+   return eval('arr[' + indexes.toString().replace(/\s/g, "").replace(/,/g, "][") + ']');
```
Comment (KostyaKov) [inline task/04-arrays-tasks.js]: To solve this task, it is more correct to process the array and not convert array to string for parsing

**Review by KostyaKov — COMMENTED** [review]
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Comment by Valeryaleshka** [comment]
I've fixed everything.
New Travis-ci build.
https://travis-ci.org/github/Valeryaleshka/js-assignments/builds/721822698

**Review by KostyaKov — COMMENTED** [review]
Closing the PR as accepted! Thanks!

**Pull request closed without merging**